File: demo-immutable-config-props-java/src/main/java/com/example/demo/configprops/AcmeProperties.java
package com.example.demo.configprops;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.util.unit.DataSize;

@ConfigurationProperties("acme")
public class AcmeProperties {

	private Duration timeout = Duration.ofSeconds(10);

	private DataSize bufferSize;

	private final Security security = new Security();

	public Duration getTimeout() {
		return this.timeout;
	}

	public void setTimeout(Duration timeout) {
		this.timeout = timeout;
	}

	public DataSize getBufferSize() {
		return this.bufferSize;
	}

	public void setBufferSize(DataSize bufferSize) {
		this.bufferSize = bufferSize;
	}

	public Security getSecurity() {
		return this.security;
	}

	public static class Security {

		private String username = "user";

		private String password;

		private List<String> roles = new ArrayList<>(Collections.singleton("USER"));

		public String getUsername() {
			return this.username;
		}

		public void setUsername(String username) {
			this.username = username;
		}

		public String getPassword() {
			return this.password;
		}

		public void setPassword(String password) {
			this.password = password;
		}

		public List<String> getRoles() {
			return this.roles;
		}

		public void setRoles(List<String> roles) {
			this.roles = roles;
		}
	}
}

File: demo-immutable-config-props-kotlin/src/main/kotlin/com/example/demo/configprops/AcmeProperties.kt
package com.example.demo.configprops

import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.context.properties.ConstructorBinding
import org.springframework.util.unit.DataSize
import java.time.Duration

@ConfigurationProperties("acme")
@ConstructorBinding
data class AcmeProperties(val timeout: Duration = Duration.ofSeconds(10),
                          val bufferSize: DataSize?,
                          val security: Security) {

    data class Security(val username: String = "user",
                        val password: String,
                        val roles: List<String> = listOf("USER"))

}